digital/Bitwise.cpp

```diff
@@ -0,0 +1,569 @@
+        _leftShift(leftShift),
+        _fcn(leftShift ? getLeftShiftFcn<Type>() : getRightShiftFcn<Type>())
+    {
+        const Pothos::DType dtype(typeid(Type), dimension);
+
+        this->setupInput(0, dtype);
+        this->setupOutput(0, dtype);
+
+        using Class = BitShift<Type>;
+
+        this->registerCall(this, POTHOS_FCN_TUPLE(Class, shiftSize));
+        this->registerCall(this, POTHOS_FCN_TUPLE(Class, setShiftSize));
+
+        this->registerProbe("shiftSize");
+        this->registerSignal("shiftSizeChanged");
+
+        // Use caller to validate input and emit signal.
+        this->setShiftSize(shiftSize);
+    }
+
+    virtual ~BitShift() = default;
+
+    size_t shiftSize() const
+    {
+        return _shiftSize;
+    }
+
+    void setShiftSize(size_t shiftSize)
+    {
+        if (shiftSize >= (sizeof(Type)*8))
+        {
+            throw Pothos::RangeException(
+                      Poco::format(
+                          "Shift size cannot be >= the number of bits (%z) in the type (%s)",
+                          (sizeof(Type)*8),
+                          Pothos::DType(typeid(Type)).toString()));
+        }
+
+        _shiftSize = shiftSize;
+        this->emitSignal("shiftSizeChanged", _shiftSize);
+    }
+
+    void work() override
+    {
+        const auto elems = this->workInfo().minElements;
+        if (0 == elems) return;
+
+        auto* inPort = this->input(0);
+        auto* outPort = this->output(0);
+
+        const auto N = elems * outPort->dtype().dimension();
+
+        _fcn(inPort->buffer(), outPort->buffer(), _shiftSize, N);
+
+        inPort->consume(elems);
+        outPort->produce(elems);
+    }
+
+private:
+    bool _leftShift;
+    size_t _shiftSize;
+    BitShiftArrayFcn<Type> _fcn;
+};
+
+//
+// Factories
+//
+
+static inline void throwExceptionForDTypeAndOperation(
+    const Pothos::DType& dtype,
+    const std::string& operation)
+{
+    throw Pothos::InvalidArgumentException(Poco::format("DType: %s, Operation: %s", dtype.toString(), operation));
+}
+
+static Pothos::Block* makeBitwiseUnaryArray(
+    const Pothos::DType& dtype,
+    const std::string& operation)
+{
+    #define BitwiseUnaryArrayFactory(T) \
+        if(Pothos::DType::fromDType(dtype, 1) == Pothos::DType(typeid(T)) && (operation == "NOT")) \
+                return new BitwiseUnaryArray<T>(dtype.dimension(), getNotFcn<T>());
+
+    BitwiseUnaryArrayFactory(std::int8_t)
+    else BitwiseUnaryArrayFactory(std::int16_t)
+    else BitwiseUnaryArrayFactory(std::int32_t)
+    else BitwiseUnaryArrayFactory(std::int64_t)
+    else BitwiseUnaryArrayFactory(std::uint8_t)
+    else BitwiseUnaryArrayFactory(std::uint16_t)
+    else BitwiseUnaryArrayFactory(std::uint32_t)
+    else BitwiseUnaryArrayFactory(std::uint64_t)
+
+    throwExceptionForDTypeAndOperation(dtype, operation);
+
+    // To stop compiler complaining, since it doesn't know the above function throws
+    return nullptr;
```

Comment:
Dont make a function `throwExceptionForDTypeAndOperation` that throws. Just make a class that inherits from `InvalidArgumentException` with the same arguments as the function. Then you just throw the class instance as usual.

Reply:
Done.